README.md
# Halite

[![Build Status](https://img.shields.io/travis/poise/halite.svg)](https://travis-ci.org/poise/halite)
[![Gem Version](https://img.shields.io/gem/v/halite.svg)](https://rubygems.org/gems/halite)
[![Coverage](https://img.shields.io/codecov/c/github/poise/halite.svg)](https://codecov.io/github/poise/halite)
[![Gemnasium](https://img.shields.io/gemnasium/poise/halite.svg)](https://gemnasium.com/poise/halite)
[![License](https://img.shields.io/badge/license-Apache_2-blue.svg)](https://www.apache.org/licenses/LICENSE-2.0)

Write as a gem, release as a cookbook.

## Quick Start

Create a gem as per normal and add a dependency on `halite`. Add
`require 'halite/rake_tasks'` to your Rakefile. Run `rake build` and the
converted cookbook will be written to `pkg/`.

All Ruby code in the gem will be converted in to `libraries/` files. You can
add cookbook-specific files by add them to a `chef/` folder in the root of the
gem.

## Why?

Developing cookbooks as gems allows using the full Ruby development ecosystem
and tooling more directly. This includes things like Simplecov for coverage
testing, YARD for documentation, and Gemnasium for dependency monitoring. For
a cookbook that is already mostly library files, this is a natural transition,
with few cookbook-specific pieces to start with. This also allows using Bundler
to manage versions instead of Berkshelf.

## Cookbook Dependencies

To add cookbook dependencies either add them to the gem requirements or use
the `halite_dependencies` metadata field:

```ruby
Gem::Specification.new do |spec|
  spec.requirements = %w{apache2 mysql}
  # or
  spec.metadata['halite_dependencies'] = 'php >= 2.0.0, chef-client'
end
```

Additionally if you gem depends on other Halite-based gems those will
automatically converted to cookbook dependencies.

## Cookbook Files

Any files under `chef/` in the gem will be written as is in to the cookbook.
For example you can add a recipe to your gem via `chef/recipes/default.rb`.

## Chef Version

By default cookbooks will be generated with `chef_version '~> 12'` to require
Chef 12.x. This can be overridden using the `halite_chef_version` metadata field:

```ruby
Gem::Specification.new do |spec|
  spec.metadata['halite_chef_version'] = '>= 12.0.0'
end
```

## Rake Tasks

The `halite/rake_tasks` module provides quick defaults. Gem name will be
auto-detected from the `.gemspec` file and the cookbook name will be based
on the gem name.

### `rake build`

The build command will convert the gem to a cookbook and write it to the `pkg/`
folder.

### Advanced Usage

You can also pass custom arguments to the Rake tasks. All parameters are
optional:

```ruby
require 'halite/rake_helper'
Halite::RakeHelper.install(
  gem_name: 'name', # Name of the gem to convert
  base: File.basename(__FILE__), # Base folder for the gem
)
```

## Berkshelf Extension

Halite includes a Berkshelf extension to pull in any gem-based cookbooks that
are available on the system.

To activate it, include the extension in your `Berksfile`:

```ruby
extension 'halite'
```

## Spec Helper

Halite includes a set of helpers for RSpec tests. You can enable them in your
`spec_helper.rb`:

```ruby
require 'halite/spec_helper'

RSpec.configure do |config|
  config.include Halite::SpecHelper
end
```

### `recipe`

Recipes to converge for the test can be defined inline on example groups:

```ruby
describe 'cookbook recipe' do
  recipe 'myrecipe'
  it { is_expected.to create_file('/myfile') }
end

describe 'inline recipe' do
  recipe do
    file '/myfile' do
      content 'mycontent'
    end
  end
  it { is_expected.to create_file('/myfile') }
end
```

### `step_into`

A resource can be added to the list to step in to via the `step_into` helper:

```ruby
describe 'mycookbook' do
  recipe 'mycookbook::server'
  step_into :mycookbook_lwrp
  it { is_expected.to ... }
end
```

### `resource` and `provider`

For testing mixin-based cookbooks, new resource and provider classes can be
declared on an example group:

```ruby
describe MyMixin do
  resource(:test_resource) do
    include MyMixin
    def foo(val=nil)
      set_or_return(:foo, val, {})
    end
  end
  provider(:test_resource) do
    def action_run
      # ...
    end
  end
  recipe do
    test_resource 'test' do
      foo 1
      action :run
    end
  end
  it { is_expected.to ... }
end
```

These helper resources and providers are only available within the scope of
recipes defined on that example group or groups nested inside it. Helper
resources are automatically `step_into`'d.

## Using a Pre-release Version of a Cookbook

When a Halite-based cookbook is released, a converted copy is generally uploaded
to [the Supermarket](https://supermarket.chef.io/). To use unreleased versions,
you need to pull in the code from git via bundler and then tell the Berkshelf
extension to convert it for you.

To grab the pre-release gem, add a line like the following to your Gemfile:

```ruby
gem 'poise-application', github: 'poise/application'
```

You will need one `gem` line for each Halite-based cookbook you want to use,
possibly including dependencies if you want to use pre-release versions of
those as well.

Next you need to use Berkshelf to convert the gem to its cookbook form:

```ruby
source 'https://supermarket.chef.io/'
extension 'halite'
cookbook 'application', gem: 'poise-application'
```

Again you will need one `cookbook` line per Halite based cookbook you want to
use. Also make sure to check the correct names for the gem and cookbook, they
may not be the same though for other Poise cookbooks they generally follow the
same pattern.

If you are using something that integrates with Berkshelf like Test-Kitchen or
ChefSpec, this is all you need to do. You could use `berks upload` to push a
converted copy of all cookbooks to a Chef Server, though running pre-release
code in production should be done with great care.

## License

Copyright 2015, Noah Kantrowitz

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Halite Changelog

## v1.8.2

* Update for newer ChefSpec.

## v1.8.1

* Fix compatibility with new Chef API for unregistering resources.

## v1.8.0

* Give recipes created using the block helper (`recipe do ... end`) a name when
  possible.
* Fix support for the Berkshelf plugin on Berks 6.1 and above.

## v1.7.0

* Allow specifying which platforms a cookbook supports via gem metadata:
  `spec.metadata['platforms'] = 'ubuntu centos'`.
* Support for automatic ChefSpec matchers in the future.

## v1.6.0

* Chef 13 compatibility.
* Structural supports for gathering development dependencies for cookbook gems.

## v1.5.0

* Set the default spec platform and version even when using the Halite runner
  API directly.
* Allow setting the chef_version constraints on a cookbook through a normal
  gem dependency on the `chef` gem.
* The default chef_version constraint has been changed from `~> 12.0` to `>= 12`
  to allow for better interaction with the upcoming Chef 13 release.
* Bump Stove dependency from 4.x to 5.x. This removes Ruby 2.0 compat. Sorry.
* No longer testing on Ruby 2.2, which happens to be entering security-maintenance
  mode today. Don't use it.

## v1.4.0

* Set a default platform and version in ChefSpec because Fauxhai is trying to
  deprecate the `chefspec` platform. This may break some tests that relied on
  the nil defaults.

## v1.3.0

* Include extended metadata with stove pushes.
* Fix `uninitialized constant Bundler::RemoteSpecification::MatchPlatform` when
  using the Berkshelf extension with ChefDK.

## v1.2.1

* Compatibility with Foodcritic 6.0. `issues_url` will be added to the generated
  `metadata.rb`. This can be set via `metadata['issues_url']` or auto-detected
  if the spec's `homepage` is set to a GitHub project.
* Compatibility with RubyGems 2.2.

## v1.2.0

* Allow passing a `Halite::Gem` object to `Halite.convert`.
* Allow disabling the stove push as part of `rake release` by setting
  `$cookbook_push=false` as an environment variable.
* Process `.foodcritic` when running `rake chef:foodcritic`.

## v1.1.0

* Support the new `chef_version` metadata field through gem metadata. Defaults
  to `~> 12` for now.
* Make spec helper resources look more like normal Chef resources when in auto
  mode.
* Add Halite's synthetic cookbooks to the cookbook compiler too, for
  include_recipe and friends.

## v1.0.13

* Additional cookbook metadata to work with Foodcritic 5.1.

## v1.0.12

* Further 12.0 fixes.

## v1.0.11

* Ensure Halite works under Chef 12.0.

## v1.0.10

* Fewer potential pitfalls when using the Halite spec helper in a non-gem
  cookbook. Still not 100%, but better.

## v1.0.9

* Additional `StubSpecifications` fixes.

## v1.0.8

* Expose `example_group` inside `SpecHelper`-created resources and providers.
* Handle `StubSpecifications` in the gem environment.

## v1.0.7

* More fixes for Chef 12.4.1.

## v1.0.6

* Minor fix for forward compatibility with 12.4.1+ and the spec helper's auto mode.

## v1.0.5

* Fix the spec helper for Chef <= 12.2.

## v1.0.4

* Fixes to work with Chef 12.4.

## v1.0.3

* Never try to do universe installs of pre-release gems in the Berkshelf extension.

## v1.0.2

* Handle converting cookbooks with pre-release version numbers and other
  non-Chef compatible components.
* Include a cookbook's changelog file in converted output.
* Handle OS X's case-insensitivity when converting misc. type files (README, etc).

## v1.0.1

* Fix issues with pre-release version numbers.

## v1.0.0

* Initial release!


lib/halite/gem.rb
#
# Copyright 2015, Noah Kantrowitz
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Because Chef 12.0 never got a backport of https://github.com/chef/chef/commit/04ba9182cda51b79630aab2918bbc6bba2d99c23
# and requiring chef/cookbook_version below can trigger the bug if singleton
# already loaded. Remove this when dropping support for 12.0.
require 'singleton'

begin
  require 'bundler' # Pull in the bundler top-level because of missing requires.
  require 'bundler/stub_specification'
rescue LoadError
  # Bundler too old.
end
require 'chef/cookbook_version'

require 'halite/dependencies'
require 'halite/error'


module Halite
  # A model for a gem/cookbook within Halite.
  #
  # @since 1.0.0
  # @example
  #   g = Halite::Gem.new('chef-mycookbook', '1.1.0')
  #   puts(g.cookbook_name) #=> mycookbook
  class Gem
    attr_reader :name

    # name can be either a string name, Gem::Dependency, or Gem::Specification
    # @param name [String, Gem::Dependency, Gem::Specification]
    def initialize(name, version=nil)
      # Allow passing a Dependency by just grabbing its spec.
      name = dependency_to_spec(name) if name.is_a?(::Gem::Dependency)
      # Stubs don't load enough data for us, grab the real spec. RIP IOPS.
      name = name.to_spec if name.is_a?(::Gem::StubSpecification) || (defined?(Bundler::StubSpecification) && name.is_a?(Bundler::StubSpecification))
      if name.is_a?(::Gem::Specification)
        raise Error.new("Cannot pass version when using an explicit specficiation") if version
        @spec = name
        @name = spec.name
      else
        @name = name
        @version = version
        raise Error.new("Gem #{name}#{version ? " v#{version}" : ''} not found") unless spec
      end
    end

    def spec
      @spec ||= dependency_to_spec(::Gem::Dependency.new(@name, ::Gem::Requirement.new(@version)))
    end

    def version
      spec.version.to_s
    end

    def cookbook_name
      if spec.metadata.include?('halite_name')
        spec.metadata['halite_name']
      else
        spec.name.gsub(/(^(chef|cookbook)[_-])|([_-](chef|cookbook))$/, '')
      end
    end

    # Version of the gem sanitized for Chef. This means no non-numeric tags and
    # only three numeric components.
    #
    # @return [String]
    def cookbook_version
      if match = version.match(/^(\d+\.\d+\.(\d+)?)/)
        match[1]
      else
        raise Halite::Error.new("Unable to parse #{version.inspect} as a Chef cookbook version")
      end
    end

    # Path to the .gemspec for this gem. This is different from
    # Gem::Specification#spec_file because the Rubygems API is shit and just
    # assumes the file layout matches normal, which is not the case with Bundler
    # and path or git sources.
    #
    # @return [String]
    def spec_file
      File.join(spec.full_gem_path, spec.name + '.gemspec')
    end

    # License header extacted from the gemspec. Suitable for inclusion in other
    # Ruby source files.
    #
    # @return [String]
    def license_header
      IO.readlines(spec_file).take_while { |line| line.strip.empty? || line.strip.start_with?('#') }.join('')
    end

    # URL to the issue tracker for this project.
    #
    # @return [String, nil]
    def issues_url
      if spec.metadata['issues_url']
        spec.metadata['issues_url']
      elsif spec.homepage =~ /^http(s)?:\/\/(www\.)?github\.com/
        spec.homepage.chomp('/') + '/issues'
      end
    end

    # Platform support to be used in the Chef metadata.
    #
    # @return [Array<Array<String>>]
    def platforms
      raw_platforms = spec.metadata.fetch('platforms', '').strip
      case raw_platforms
      when ''
        []
      when 'any', 'all', '*'
        # Based on `ls lib/fauxhai/platforms  | xargs echo`.
        %w{aix amazon arch centos chefspec debian dragonfly4 fedora freebsd gentoo
           ios_xr mac_os_x nexus omnios openbsd opensuse oracle raspbian redhat
           slackware smartos solaris2 suse ubuntu windows}.map {|p| [p] }
      when /,/
        # Comma split mode. String looks like "name, name constraint, name constraint"
        raw_platforms.split(/\s*,\s*/).map {|p| p.split(/\s+/, 2) }
      else
        # Whitepace split mode, assume no constraints.
        raw_platforms.split(/\s+/).map {|p| [p] }
      end
    end

    # Iterate over all the files in the gem, with an optional prefix. Each
    # element in the iterable will be [full_path, relative_path], where
    # relative_path is relative to the prefix or gem path.
    #
    # @param prefix_paths [String, Array<String>, nil] Option prefix paths.
    # @param block [Proc] Callable for iteration.
    # @return [Array<Array<String>>]
    # @example
    #   gem_data.each_file do |full_path, rel_path|
    #     # ...
    #   end
    def each_file(prefix_paths=nil, &block)
      globs = if prefix_paths
        Array(prefix_paths).map {|path| File.join(spec.full_gem_path, path) }
      else
        [spec.full_gem_path]
      end
      [].tap do |files|
        globs.each do |glob|
          Dir[File.join(glob, '**', '*')].each do |path|
            next unless File.file?(path)
            val = [path, path[glob.length+1..-1]]
            block.call(*val) if block
            files << val
          end
        end
        # Make sure the order is stable for my tests. Probably overkill, I think
        # Dir#[] sorts already.
        files.sort!
      end
    end

    # Special case of the {#each_file} the gem's require paths.
    #
    # @param block [Proc] Callable for iteration.
    # @return [Array<Array<String>>]
    def each_library_file(&block)
      each_file(spec.require_paths, &block)
    end

    def cookbook_dependencies(development: false)
      Dependencies.extract(spec, development: development)
    end

    # Is this gem really a cookbook? (anything that depends directly on halite and doesn't have the ignore flag)
    def is_halite_cookbook?
      spec.dependencies.any? {|subdep| subdep.name == 'halite'} && !spec.metadata.include?('halite_ignore')
    end

    # Create a Chef::CookbookVersion object that represents this gem. This can
    # be injected in to Chef to simulate the cookbook being available.
    #
    # @return [Chef::CookbookVersion]
    # @example
    #   run_context.cookbook_collection[gem.cookbook_name] = gem.as_cookbook_version
    def as_cookbook_version
      # Put this in a local variable for a closure below.
      path = spec.full_gem_path
      Chef::CookbookVersion.new(cookbook_name, File.join(path, 'chef')).tap do |c|
        # Use CookbookVersion#files_for as a feature test for ManifestV2. This
        # can be changed to ::Gem::Requirement.create('>= 13').satisfied_by?(::Gem::Version.create(Chef::VERSION))
        # once https://github.com/chef/chef/pull/5929 is merged.
        if defined?(c.files_for)
          c.all_files = each_file('chef').map(&:first)
        else
          c.attribute_filenames = each_file('chef/attributes').map(&:first)
          c.file_filenames = each_file('chef/files').map(&:first)
          c.recipe_filenames = each_file('chef/recipes').map(&:first)
          c.template_filenames = each_file('chef/templates').map(&:first)
        end
        # Haxx, rewire the filevendor for this cookbook to look up in our folder.
        # This is touching two different internal interfaces, but ¯\_(ツ)_/¯
        c.send(:file_vendor).define_singleton_method(:get_filename) do |filename|
          File.join(path, 'chef', filename)
        end
        # Store the true root for use in other tools.
        c.define_singleton_method(:halite_root) { path }
      end
    end

    # Search for a file like README.md or LICENSE.txt in the gem.
    #
    # @param name [String] Basename to search for.
    # @return [String, Array<String>]
    # @example
    #   gem.misc_file('Readme') => /path/to/readme.txt
    def find_misc_path(name)
      [name, name.upcase, name.downcase].each do |base|
        ['.md', '', '.txt', '.html'].each do |suffix|
          path = File.join(spec.full_gem_path, base+suffix)
          return path if File.exist?(path) && Dir.entries(File.dirname(path)).include?(File.basename(path))
        end
      end
      # Didn't find anything
      nil
    end

    # Figure out which version of Chef this cookbook requires. Returns an array
    # of Gem::Requirement-style string like `~> 12.0`.
    #
    # @return [Array<String>]
    def chef_version_requirement
      if spec.metadata['halite_chef_version']
        # Manually overridden by gem metadata, use that.
        [spec.metadata['halite_chef_version']]
      elsif dep = spec.dependencies.find {|inner_dep| inner_dep.name == 'chef' && !inner_dep.requirement.none? }
        # Parse through the dependencies looking for something like `spec.add_dependency 'chef', '>= 12.1''`.
        dep.requirement.as_list
      else
        ['>= 12']
      end
    end

    private

    # Find a spec given a dependency.
    #
    # @since 1.0.1
    # @param dep [Gem::Dependency] Dependency to solve.
    # @return [Gem::Specificiation]
    def dependency_to_spec(dep)
      # #to_spec doesn't allow prereleases unless the requirement is
      # for a prerelease. Just use the last valid spec if possible.
      spec = dep.to_spec || dep.to_specs.last
      raise Error.new("Cannot find a gem to satisfy #{dep}") unless spec
      spec
    rescue ::Gem::LoadError => ex
      raise Error.new("Cannot find a gem to satisfy #{dep}: #{ex}")
    end
  end
end
